blog/_src/posts/2019-08-14-lexical-and-dynamic-scope.md

```diff
@@ -0,0 +1,246 @@
+    Title: Lexical and Dynamic Scope
+    Date: 2019-08-14T13:59:21
+    Tags: scope, definitions, history, by Ming-Ho Yee
+
+This all started with a simple question about the R programming language: _is R
+lexically or dynamically scoped?_
+
+To answer that question, we need to understand what _scope_ is, along with
+_lexical scope_ and _dynamic scope_.
+
+In this blog post, I'd like to explain the differences between lexical scope
+and dynamic scope, and also explore some of the history behind those ideas. In
+a subsequent post, I'll discuss scoping in R and why it can be confusing.
+
+<!-- more -->
+
+## What is scope?
+
+_Scope_ refers to the places in a program where a variable is visible and can
+be referenced.
+
+An interesting situation is when a function has free variables. Consider the
+example below:
+
+```r
+x <- 1
+f <- function(a) x + a
+g <- function() {
+  x <- 2
+  f(0)
+}
+g() # what does this return?
+```
+
+In this snippet, we define a variable `x` at the top level and two functions
+`f` and `g`. The body of `f` uses the parameter `a`, but also the free variable
+`x`. On line 7, we call `g`, which in turn calls `f` and returns that result.
+
+What value does `g` return when it is called? On line 2, what value should the
+free variable `x` be bound to?
+
+
+### Lexical scoping
+
+Under _lexical scoping_ (also known as _static scoping_), the scope of a
+variable is determined by the lexical (_i.e._, textual) structure of a program.
+
+In the example above, the definition of `x` on line 1 creates a scope that
+starts after its definition, and extends _into_ the bodies of `f` and `g`.
+However, the second definition of `x` on line 4 creates a new scope that (1)
+shadows the previous definition of `x`, and (2) does not extend into the call
+`f(0)` on line 5. Looking at this from another direction, the use of `x` on
+line 2 is within the scope created by the definition on line 1, and thus refers
+to that definition.
+
+Therefore, under lexical scoping, the example program returns `1`.
+
+Most programming languages we use today are lexically scoped. Intuitively, a
+human (or compiler) can determine the scope of a variable by just examining the
+source code of a program. In other words, a compiler can determine which
+_definition_ each variable refers to---but it may not be able to determine the
+_values_ of each variable.
+
+
+### Dynamic scoping
+
+Under _dynamic scoping_, a variable is bound to the most recent value assigned
```

Comment:
I find this talk of assignment a bit confusing. It's evocative of mutable state, which seems orthogonal to scoping.

Reply:
This is tough. In the previous draft, I explained scoping from a "bottom up" perspective, i.e. variable lookup. But that bothered me, so I switched to a "top down," from the perspective of a variable definition (or assignment). It's still a bit weird, but I think it better fits the "scope is the places in a program where a variable reference is valid" story.

Later on, I try to explain why this definition of dynamic scope seems weird. I can try to clarify a little bit, a little earlier.